main.py

```diff
@@ -44,55 +37,66 @@ class MainApp(QtWidgets.QMainWindow, Ui_MainWindow):
         self.longitude = None
-        self.lastLatitude = None
-        self.lastLongitude = None
-        self.x = None
-        self.y = None
 
         QtWidgets.QMainWindow.__init__(self)
         Ui_MainWindow.__init__(self)
         self.setupUi(self)
 
+        # Attach functions for buttons
         self.sendButton.clicked.connect(self.sendButtonPressed)
         self.commandEdit.returnPressed.connect(self.sendButtonPressed)
-        self.actionSave.triggered.connect(self.data.save)
+        self.actionSave.setShortcut("Ctrl+S")
+        self.actionSave.triggered.connect(self.saveFile)
 
         self.StatusButton.clicked.connect(lambda _: self.sendCommand("status"))
         self.ArmButton.clicked.connect(lambda _: self.sendCommand("arm"))
         self.HaloButton.clicked.connect(lambda _: self.sendCommand("halo"))
 
-        markerpath = os.path.join(local, "marker.png")
+        # Map UI attributes
+        self.zoom = 20
+        self.radius = 0.1
+
+        markerpath = os.path.join(LOCAL, "marker.png")
         # TODO: imresize removed in latest scipy since it's a duplicate from "Pillow". Update and replace.
         self.marker = imresize(plt.imread(markerpath), (12, 12))
         # self.plotMap(51.852667, -111.646972, self.radius, self.zoom)
         # self.plotMap(49.266430, -123.252162, self.radius, self.zoom)
 
-        idSet = set(RocketData.chartoname.keys())
         self.printToConsole("Starting Connection")
-        self.SThread = SerialThread.SThread(connection, idSet)
-        self.SThread.sig_received.connect(self.receiveData)
-        self.sig_send.connect(self.SThread.queueMessage)
-        self.SThread.sig_print.connect(self.printToConsole)
-        self.SThread.start()
 
+        # Init and connection of ReadThread
+        self.ReadThread = ReadThread.ReadThread(self.connection, self.data)
+        self.ReadThread.sig_received.connect(self.receiveData)
+        self.ReadThread.sig_print.connect(self.printToConsole)
+        self.ReadThread.start()
+
+        # Init and connection of SendThread
+        self.SendThread = SendThread.SendThread(self.connection)
+        self.sig_send.connect(self.SendThread.queueMessage)
+        self.SendThread.sig_print.connect(self.printToConsole)
+        self.SendThread.start()
+
+        # Mapping thread init and start
         thread = threading.Thread(target=self.threadLoop, daemon=True)
-        thread.start()
+        #thread.start()
+
+    def closeEvent(self, event):
+        self.connection.shutDown()
+        print("Saving...")
+        self.data.save(os.path.join(LOCAL, "finalsave_" + str(int(time.time())) + ".csv"))
+        print("Saved!")
 
-    # Possible doc: receives from the serial thread loop of subpackets
-    def receiveData(self, dataBundle):  # TODO: ARE WE SURE THIS IS THREAD SAFE? USE QUEUE OR PUT IN SERIAL THREAD
-        self.data.addBundle(dataBundle)
```

Comment:
@ADSteele916, main.py is no longer adds data to RocketData. This means that main.py isn't responsible for anything other than the UI now (aside from some set-up at init). Hopefully this can help you with your tests as running without a UI should be much easier now.

Reply:
That will be welcome. 👍